### ISSUE TITLE
Investigate variadic templates to decrease code size

### ISSUE THREAD

**Issue by citibeth** [blitzpp/blitz#9]
Peter Boyle wrote:

Just a small comment — I’m working on a distributed MPP cartesian grid code.

github.com/paboyle/Grid

Certainly not pitching it as it is niche, focussed on particle physics, but I found C++11 does enable MUCH smaller more
compact expression template engines ~ 250 LOC. Something I learned from Antonin Portelli, CC’d. The corresponding C++98
US  QDP++ package was more like 100k LOC using PETE.

So I agree that there is VERY much scope for a rewrite/update using features that were not originally available.

**Comment by lutorm** [comment]
I agree that with C++11 much ugliness could be removed. Just being able to use "auto" for types would probably cut the code in half... ;-)

Two points, however:
1. Not everyone can use C++11, so before we drop support for C++98 I think it's worth thinking about things that can be improved/fixed in the current code base.
2. We should not rewrite the code just because we can. It would be a big project, and we'd go from a fairly well-tested and user-proven code base to a new one, with new and unknown bugs. If we decide to start working on this, we should make sure there's enough people that are committed to seeing it through that we don't end up with something half-completed.

**Comment by citibeth** [comment]
I agree, moving to variadic templates is low priority at this point.  And a
C++98 compatible version (1.0) needs to be released and supported before
C++11-only versions (2.0) are produced.

**Comment by AnderOne** [comment]
This PR (#117) demonstrates case when variadic templates may be helpful for a checking function arguments in compile time.
We can't check all possible mixtures of argument types without them.

For example:
```
A(Range, Range)
A(int, Range)
A(Range, int)
A(int, int)
```

For `Array<T, N>` we will get 2^{N} different specializations!
